HDDS-8605. Implement the ability to update the ServiceInfo object with the new rootCA

fapifta opened the pull request:
## What changes were proposed in this pull request?

During the rotation of the rootCA certificate, we need to update the cached PEM format representation of the CA certificates in Ozone Manager, so that via the ServiceInfoEx object we are able to provide those to the clients.
In this PR, I have separated a class called ServiceInfoProvider, into which I have pushed out all the relevant code.
It attaches a listener to the OM's certificate client that manages the polling for the rootCA certificate, and then notifies this provider to update its cached data.
OM every time creates a ServiceInfoEx object once it is requested via the OzoneManagerProtocol, so does the provider.

The provider is needed to separate away the logic and to be able to test the handling of the renew separately and effectively.

In order to utilize some pre-written utilities in the test for the Provider, I have moved the relevant methods from TestInterSCMProtocolService to a test utility class under hdds-common. With that I had to move the HDDSKeyGenerator utility class in the production code as well, so that it is available for the test utilities. Both of these are general utilities, so they are rightfully placed in hdds-common I believe.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8605

## How was this patch tested?
Added UT for the new functionality.

ChenSammi commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/HddsConfigKeys.java:
```diff
@@ -216,6 +216,10 @@ public final class HddsConfigKeys {
   public static final String HDDS_X509_CA_ROTATION_TIME_OF_DAY_DEFAULT =
       "02:00:00";
+  public static final String HDDS_X509_ROOTCA_CLIENT_POLLING_FREQUENCY =
+      "hdds.x509.rootca.client.polling.frequency";
```
I would suggest user "interval" instead of "frequency" here, to be consistent with existing other properties.

fapifta replied:
@Galsza I believe this comment is for your patch at #4961 ;)

Galsza replied:
Yes, that's a better word, I'm going to fix it

adoroszlai commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/CertificateClient.java:
```diff
@@ -126,6 +130,30 @@ X509Certificate getCertificate(String certSerialId)
+   * @return either the rootCA certificate, or the sub-CA certificate
+   *          of the node that served as the rootCA during bootstrap.
+   */
+  default String getSingleCACertificatePEM() throws SCMSecurityException {
+    return CertificateCodec.getPEMEncodedString(getSingleCACertificate());
+  }
```
Since this is only used by `ServiceInfoProvider`, which already handles PEM encoding for the CA list, I think this method should be omitted.

fapifta replied:
Yes, this makes testing a bit more complex though I think you are right, a simpler API for the CertificateClient interface is more important, I haven't realized this one before, thank you for catching it.

adoroszlai replied:
There are a few more tests that need to be updated:

https://github.com/fapifta/hadoop-ozone/actions/runs/5459036651/jobs/9934647451#step:5:3792
https://github.com/fapifta/hadoop-ozone/actions/runs/5459036651/jobs/9934647503#step:5:3764
https://github.com/fapifta/hadoop-ozone/actions/runs/5459036651/jobs/9934647503#step:5:3810

Sorry for the trouble. :)

fapifta replied:
Yes, I have adjusted the code to overcome these.

adoroszlai commented on hadoop-ozone/ozone-manager/src/test/java/org/apache/hadoop/ozone/om/TestServiceInfoProvider.java:
```diff
@@ -0,0 +1,138 @@
+import org.apache.hadoop.ozone.om.helpers.ServiceInfo;
+import org.apache.hadoop.ozone.om.helpers.ServiceInfoEx;
+import org.apache.hadoop.ozone.om.protocol.OzoneManagerProtocol;
+import org.junit.Test;
```
It would be nice to implement new tests using JUnit5.

fapifta replied:
You are right, I have changed it. With that I realized we do not need the MockitoJunitRunner either.

adoroszlai commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ServiceInfoProvider.java:
```diff
@@ -0,0 +1,102 @@
+    } else {
+      this.certificateClient = null;
+      caCertPem = null;
+      caCertPemList = null;
```
Please use `emptyList()` instead of `null` for `caCertPemList` (in all branches).

```
om1_1       | java.lang.NullPointerException
om1_1       |   at org.apache.hadoop.ozone.protocolPB.OzoneManagerRequestHandler.getServiceList(OzoneManagerRequestHandler.java:855)
```

fapifta replied:
Nice catch, thank you for tracking this one, I have missed the original initialization of the caCertPemList field. Fixed.

adoroszlai commented on hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/security/x509/package-info.java:
```diff
@@ -0,0 +1,99 @@
+ * framework for HDDS.
+ */
+package org.apache.hadoop.hdds.security.x509;
+/*
+
+Architecture of Certificate Infrastructure for SCM.
+====================================================
```
I think this should be a javadoc comment on the package, otherwise most people will never see it.  At least I didn't know about it until now. ;)

Also, I don't think we should duplicate this description in yet another place, it's already present in:

```
hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/x509/package-info.java
hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/security/x509/package-info.java
```

It should be in `common/src/main` only, and the others should refer to it.

fapifta replied:
Actually this description is somewhat the tldr of the original design document in HDDS-4. Parts of it is obsolete, some parts were never implemented. I decided to remove this, as with the current and recent changes we'd better write a solid documentation page rather than trying to summarize this in the APIDoc I believe.

I removed the duplicated package-info file form the hdds/framework module.

adoroszlai reviewed (commented):
Thanks @fapifta for the patch.

fapifta commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/OzoneManager.java:
```diff
@@ -640,6 +638,16 @@ private OzoneManager(OzoneConfiguration conf, StartupOption startupOption)
+          } catch (Exception e) {
+            throw new RuntimeException(e);
+          }
+        });
```
Ah, yes, I forget about that class... Thank you for the suggestion!

Galsza commented on hadoop-hdds/common/src/main/java/org/apache/hadoop/hdds/security/x509/certificate/client/CertificateClient.java:
```diff
@@ -126,6 +130,30 @@ X509Certificate getCertificate(String certSerialId)
+   * @return either the rootCA certificate, or the sub-CA certificate
+   *          of the node that served as the rootCA during bootstrap.
+   */
+  default X509Certificate getSingleCACertificate() {
```
This method could possibly used in other places, (ReloadingTrustManager is what I know by heart, but it's going to change with Sammi's change), could you check once we get there, if there are other places we could use this?

fapifta replied:
Actually after the new changes, I realized that this method is not needed anymore, and I should use the new methods that we have in the certificate client to get the list of rootCA certificates. I have reworked the relevant code, and related tests.

Galsza commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ServiceInfoProvider.java:
```diff
@@ -0,0 +1,102 @@
+      onRootCAChange() {
+    return certs -> {
+      try {
+        caCertPem = certificateClient.getSingleCACertificatePEM();
```
I think there is a race condition here with the certificateClient saving its own new rootCaCertificate. On root ca rotation the client will schedule a task to send a CSR and only after receiving the signed certificate will it save and update the new RootCA/SubCA. However when the ServiceInfoProvider asks for it it still might not have reached that point and therefore would receive the old certificate data. (And this notification won't repeat because we already think that the handlers successfully rotated the root ca)

I have a suggestion which could resolve this issue: instead of registering this to the rootCaRotationListeners we could register it to just the CertificateNotation receivers. They will get notified each cert rotation including the client renewing their known root ca.

fapifta replied:
The thing is that, it is a race condition, but the problem is not the race...

The problem is more of a question instead, around what we want to do with 1.1 or older clients, as all 1.2 or newer client versions work based off of the caCertPemList instead of the caCertPem value.
So newer clients rightfully get the list of the old and the new rootCA certificate, and will work with the cluster fine during the rotation, whichever root of trust the certificates of cluster services have old or new.

But with older clients we have a similar tradeoff as we had once we have introduce SCM HA. With SCM HA old clients got the sub-CA certificate that OM has read from its filesystem as the signer of its certificate, and if the post 1.2 DN that the pre 1.2 DN trying to connect to has its certificate signed by the same cert, then the client was able to read from the DN. Now in our case, we again need a decision, either we give back the old root CA certificate, which is returned by the certClient at this point, or we return the new certificate.

Based on your question I am happy that I thought this through more, and I think we should return the new rootCA certificate here. With that, we will have a glitch in old clients, as we don't know how many DNs have already renewed their certificates with the new rootCA and its subCAs, but for sure the number will be all DNs eventually, while less and less DNs will use their old certificates. So I need to change this code accordingly. Thank you for pointing this out!

Galsza reviewed (commented):
Thanks @fapifta for working on this. I have left 2 comments, one of them is minor, but I'm curious about your opinion on the suggestion of avoiding the race condition between the client's cert renewal and the ServiceInfoProvider's. I hope I didn't misunderstand anything.

fapifta commented:
I think all the review comments were addressed, @Galsza @ChenSammi @adoroszlai please take a look once you have the time.

I have rebased the patch onto the tip of the master branch, as it has all the related commits. During the process I squashed my commits that were added to address the review comments, but in the meantime, I think I have improved the readability of the code. Let me know if you see any further things to address.
Thank you!

Galsza commented:
@fapifta Thank you Pifta, there are a few elegant solutions here that I could learn of. LGTM +1

ChenSammi commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ServiceInfoProvider.java:
```diff
@@ -0,0 +1,145 @@
+  }
+
+  private Set<X509Certificate> getCACertificates() {
+    return !certClient.getAllRootCaCerts().isEmpty()
```
Please avoid call certClient.getAllRootCaCerts() twice.

fapifta replied:
It does not have any significant impact, the getter itself provides an unmodifiable set, and that is basically just a wrapper around the internal data structure inside the CertificateClient.
I am fine with both approach so I have updated the code.

ChenSammi commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ServiceInfoProvider.java:
```diff
@@ -0,0 +1,145 @@
+      boolean skipInitializationForTesting
+  ) throws IOException {
+    this.om = om;
+    if (new SecurityConfig(config).isSecurityEnabled()
```
"certClient != null" will have the same effect as "new SecurityConfig(config).isSecurityEnabled()" and avoid create a new SecurityConfig object.

fapifta replied:
I realized that the configuration itself was needed just for this one. So instead I pass on a SecurityConfig object from OzoneManager directly. I know the null check is essentially the same effect, but this way the code is better expressing what it does, and why.

ChenSammi commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ServiceInfoProvider.java:
```diff
@@ -0,0 +1,145 @@
+    try {
+      return toPEMEncodedString(cert);
+    } catch (SCMSecurityException e) {
+      return null;
```
@fapifta , I think we should not hide the exception.

fapifta replied:
Hiding this is not really a problem. The following logic is applied on the client side: if there are a list of CA certs, then the list is being used, if there is no element in the list or the list is null, then the sole caCertPEM is being used. This logic is there to ensure compatibility between the new (1.2+) client and an older server side, so we need to keep that semantic.

If any/all certificates in the list fails to translate to PEM format, then we have a second chance.

I have added an error level log message to sign that there is a problem, but chances are that this will never happen, as we get the in memory X509Certificate already from a PEM formatted string, so we should fail earlier if there is a problem with this conversion.

ChenSammi replied:
@fapifta , thanks for the explanation. 
It cannot purely sure a checked type exception will not happen.  Let's not make that assumption. 
Besides, an originally empty caCertPEMList, and an empty caCertPEMList because of certificate convert failure are two different cases.  We need to distinguish them. Not sure if the backward comparability handling on client side can cover them both. Even if it can, it's not a recommended behavior.

fapifta replied:
Looking at it once more I relaized, that this local optimization does not really mean a thing, as other methods might throw an exception and still fail early. So instead of returning null, the exception is re-thrown as runtimeexception.

ChenSammi commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ServiceInfoProvider.java:
```diff
@@ -0,0 +1,145 @@
+
+  private X509Certificate newestOf(Collection<X509Certificate> certs) {
+    return certs.stream()
+        .max((c1, c2) -> expiry(c1).compareTo(expiry(c2)))
```
Please use the inline statement directly, instead of calling a one statement function.

fapifta replied:
JIT would inline the method call anyways, and probably it would do more. Though this note was useful because it made me check what I was doing, and realize that I have a way better solution to solve the readability problem I wanted to solve with the expiry method.

ChenSammi commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ServiceInfoProvider.java:
```diff
@@ -0,0 +1,145 @@
+        caCertPemList = toPEMEncodedStrings(certs);
+        returnedFuture.complete(null);
+      } catch (Exception e) {
+        LOG.info("Unable to refresh cached PEM formatted CA " +
```
info -> warn

fapifta replied:
done.

ChenSammi commented on hadoop-ozone/ozone-manager/src/main/java/org/apache/hadoop/ozone/om/ServiceInfoProvider.java:
```diff
@@ -0,0 +1,145 @@
+  }
+
+  public ServiceInfoEx provide() throws IOException {
+    return new ServiceInfoEx(om.getServiceList(), caCertPem, caCertPemList);
```
Need a way to synchronize the fetch and update of caCertPem and caCertPermList.

fapifta replied:
The easiest seems to synchronize on this when the values are updated in onRootCAChange, and then assign these to variables in a synchronized block so that we can provide a copy to the ServiceInfoEx constructor, and we do not need to care about further synchronization if I think it correclty. Adjusted the code.

ChenSammi commented:
Hi @fapifta , thanks for working on this. Could you add some file put/get operation in robot test, after the root certificate is rotated and first root certificate is expired?  You can leverage test-root-ca-rotation.sh

fapifta commented on hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/security/x509/CertificateTestUtils.java:
```diff
@@ -0,0 +1,127 @@
+    final X509v3CertificateBuilder certificateBuilder =
+        new JcaX509v3CertificateBuilder(
+            x500Name,
+            BigInteger.valueOf(keys.getPublic().hashCode()),
```
As this is a utility method, I could design it only for the use case I have at hand. I am adjusting the documentation, but I think we should not overengineer it here, once someone needs some other functionality, we can add it at that point in time either via an overload or via a new parameter and adjusted code where it is used.

fapifta commented:
Thank you for the additional reviews @ChenSammi @Galsza. I tried to address all your comments properly, if you can take an other look, I would be glad.

This PR in the current stage relies on #5050 and #5052 to be added, until that the ozonesecure and ozonesecure-ha tests are expected to fail.

fapifta commented:
After I was able to merge #5050 and #5052 I have rebased this branch on master. Pending CI and secondary reviews.

ChenSammi commented on hadoop-ozone/dist/src/main/compose/ozonesecure-ha/test-root-ca-rotation.sh:
```diff
@@ -67,6 +68,7 @@ wait_for_execute_command scm4.org 240 "ozone admin cert info 4"
 
 #transfer leader to scm4.org
 execute_robot_test scm4.org -v "TARGET_SCM:scm4.org" scmha/scm-leader-transfer.robot
+execute_robot_test scm4.org -v PREFIX:"rootca" certrotation/root-ca-rotation-client-checks.robot
```
rootca -> rootca2

I checked the failed robot test.  After leader is transferred to scm4.org,  leader OM will fail to allocate block for key upload since the three OM doesn't know scm4.org. We can move this root-ca-rotation-client-checks.robot test ahead of leader transfer.

fapifta replied:
updates, thank you for checking into this.

ChenSammi commented on hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/security/x509/CertificateTestUtils.java:
```diff
@@ -0,0 +1,164 @@
+   */
+  public static X509Certificate createSelfSignedCert(
+      KeyPair keys, String commonName
+  ) throws Exception {
```
I would like to discuss the function declare code style here.  It's better to keep the consistent style with other majority codes, putting as many as possible in one line, 

```
public static X509Certificate createSelfSignedCert(
      KeyPair keys, String commonName) throws Exception {
```

fapifta replied:
I also updated the constructor of ServiceInfoProvider to conform this style.

ChenSammi commented on hadoop-hdds/common/src/test/java/org/apache/hadoop/hdds/security/x509/CertificateTestUtils.java:
```diff
@@ -0,0 +1,164 @@
+   */
+  public static X509Certificate createSelfSignedCert(
+      KeyPair keys, String commonName, Duration expiresIn
+  ) throws Exception {
```
same as above comment.

ChenSammi commented:
@fapifta , thanks for update the patch. I have left four new comments, including the one under toPEMEncodedStringUnsafe.

fapifta commented:
Thank you for the review, I have pushed changes to address your comments @ChenSammi, thank you for the continous reviews!

fapifta commented:
Looking at it further after the change in ServiceInfoProvider.toPEMEncodedStringUnsafe, the method felt unnecessary, so I did some more code changes around that, and removed the method.
@ChenSammi please take a look once you have time.

ChenSammi commented:
@fapifta , I checked the failed UT, looks like it's because "hdds.x509.ca.rotation.enabled" default is false now,  so RootCaRotationPoller is not instantiated.

fapifta commented:
Oh well... thank you @ChenSammi to take a look at, I am verifying the fix for this quickly on the failed tests, and push the fix.

ChenSammi commented:
Thanks @fapifta ,  LGTM +1.
